### PR TITLE
[SPARK-40752][SQL] Migrate type check failures of misc expressions onto error classes

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
This pr replaces TypeCheckFailure by DataTypeMismatch in type checks in the misc expressions, includes:
1. Coalesce [nullExpressions.scala#L60](https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala#L60)
2. SortOrder [SortOrder.scala#L75](https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala#L75)
3. UnwrapUDT [UnwrapUDT.scala#L36](https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/UnwrapUDT.scala#L36)
4. ParseUrl [urlExpressions.scala#L185](https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala#L185)
5. XPathExtract [xpath.scala#L45](https://github.com/apache/spark/blob/a241256ed0778005245253fb147db8a16105f75c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/xpath.scala#L45)

### Why are the changes needed?
Migration onto error classes unifies Spark SQL error messages.

### Does this PR introduce _any_ user-facing change?
Yes. The PR changes user-facing error messages.

### How was this patch tested?
1. Add new UT
2. Update existed UT
3. Pass GA

### CODE REVIEW

**Comment by MaxGekk** [comment]
@panbingkun Please, fix the coding style issue:
```
[error] /home/runner/work/spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/xpath.scala:21:82: Unused import
[error] import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.{DataTypeMismatch, TypeCheckFailure}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala [hunk 1]
```diff
@@ -72,7 +74,11 @@ case class SortOrder(
     if (RowOrdering.isOrderable(dataType)) {
       TypeCheckResult.TypeCheckSuccess
     } else {
-      TypeCheckResult.TypeCheckFailure(s"cannot sort data type ${dataType.catalogString}")
+      DataTypeMismatch(
+        errorSubClass = "UNSUPPORTED_INPUT_TYPE",
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala:78]: Could you re-use the existing error class `INVALID_ORDERING_TYPE`. Just call `TypeUtils.checkForOrderingExpr(...)` here.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Comment by MaxGekk** [comment]
@panbingkun Please, remove duplicate items in PR's description:
```
- 1.Add new UT
- 2.Update existed UT
- 3.Pass GA.
```
to
```
1. Add new UT
2. Update existed UT
3. Pass GA.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala (hunk 1 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala]: Done

**Comment by panbingkun** [comment]
> @panbingkun This pr **replace** TypeCheckFailure -> This pr **replaces** TypeCheckFailure

Done.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Pull request closed without merging**